cpp/arcticdb/storage/s3/detail-inl.hpp

```diff
@@ -44,50 +44,47 @@ namespace s3 {
+                raise<ErrorCode::E_PERMISSION>(fmt::format("Permission error: S3Error#{} {}: {}",
+                                                           int(err.GetErrorType()),
+                                                           err.GetExceptionName().c_str(),
+                                                           err.GetMessage().c_str()));
+            }
+
+            if(err.ShouldRetry()) {
+                raise<ErrorCode::E_S3_RETRYABLE>(fmt::format("Retry-able error: S3Error#{} {}: {}",
+                                                           int(err.GetErrorType()),
+                                                           err.GetExceptionName().c_str(),
+                                                           err.GetMessage().c_str()));
             }
-        private:
-            std::string message;
-        };
 
-        inline void raise_s3_exception(const Aws::S3::S3Error& err){
-            std::string error_message;
             // We create a more detailed error explanation in case of NETWORK_CONNECTION errors to remedy #880.
-            if (err.GetErrorType() == Aws::S3::S3Errors::NETWORK_CONNECTION){
-                error_message = fmt::format("Got an unexpected network error: {}: {}. "
+            if (type == Aws::S3::S3Errors::NETWORK_CONNECTION) {
+                error_message = fmt::format("Unexpected network error: S3Error#{}: {} {} "
                                             "This could be due to a connectivity issue or too many open Arctic instances. "
                                             "Having more than one open Arctic instance is not advised, you should reuse them. "
                                             "If you absolutely need many open Arctic instances, consider increasing `ulimit -n`.",
+                                            int(err.GetErrorType()),
                                             err.GetExceptionName().c_str(),
                                             err.GetMessage().c_str());
             }
             else {
-                error_message = fmt::format("Got unexpected error: '{}' {}: {}",
+                error_message = fmt::format("Unexpected error: S3Error#{} {}: {}",
                                             int(err.GetErrorType()),
                                             err.GetExceptionName().c_str(),
                                             err.GetMessage().c_str());
             }
 
             log::storage().error(error_message);
-            throw UnexpectedS3ErrorException(error_message);
-        }
-
-        inline bool is_expected_error_type(Aws::S3::S3Errors err) {
-            return err == Aws::S3::S3Errors::NO_SUCH_KEY
-                   || err == Aws::S3::S3Errors::NO_SUCH_BUCKET
-                   || err == Aws::S3::S3Errors::INVALID_ACCESS_KEY_ID
-                   || err == Aws::S3::S3Errors::ACCESS_DENIED
```

Comment:
So previously it would throw "key doesn't exist", but now we explicitly will throw a permissions error. 👍